File: apps/ios/Packages/StoryArcKit/Sources/PlayerFeature/PlayerArtwork.swift
public import SwiftUI

public import StoryArcCore

internal import DesignSystem

/// The player's artwork: the shared coverless well, at the player's own shape.
///
/// `audio-playback`, *A publication with no cover*: the player "draws the same coverless
/// treatment every other surface draws … rather than a generic glyph", and "the system's own
/// media controls get that same artwork, because a lock screen showing a headphones symbol is
/// the one place a listener looks for an hour".
///
/// **It now literally is the same view.** `FullPlayerView` drew `Image(systemName:
/// "headphones")` with a comment claiming it was "the same placeholder the library draws";
/// that was replaced by a title set into a rounded rectangle here, which was a second
/// implementation of the treatment rather than the treatment — the compromise
/// `audiobooks-and-playback` §4.4b recorded, because ``CoverlessWell`` was in `LibraryFeature`
/// and `Package.swift` forbids one feature depending on another. The well is in
/// `DesignSystem` now, which `PlayerFeature` already depends on, so this file is a shape and
/// nothing else.
///
/// The glyph is chosen from the format rather than fixed, which is what makes the spec's
/// "rather than a generic glyph" true of a read-aloud EPUB as well as of an M4B: the two are
/// one player and two different things to look at.
///
/// **A square rather than 2:3, and that is not a second treatment.** An audiobook's artwork is
/// square everywhere a listener has seen one, and the rendered image below is what a lock
/// screen and a car display are handed. The well fills whatever shape it is given and sizes
/// its glyph off that shape, so the same view is right at 146 pt in a grid cell and at 320 pt
/// here.
public struct PlayerArtwork: View {
    private let format: PublicationFormat

    public init(format: PublicationFormat) {
        self.format = format
    }

    public var body: some View {
        CoverlessWell(format: format)
            .aspectRatio(1, contentMode: .fit)
            .clipShape(.rect(cornerRadius: StoryArcRadius.lg, style: .continuous))
    }
}

/// The same artwork as pixels, for the system's own media controls.
///
/// `Playback` has no SwiftUI and must not — `Formats` depends on it, and a parser has no
/// business linking a design system — so the picture is drawn here, where the treatment lives,
/// and carried to `NowPlaying` through ``PlayerCentre/onArtwork`` as the one thing both a
/// SwiftUI view and an `MPMediaItemArtwork` can hold.
///
/// **Rendered from the same view the player draws**, rather than re-drawn in Core Graphics: a
/// second implementation of the treatment is a second thing to keep in step, and the spec asks
/// for "that same artwork" rather than a similar one.
///
/// The title is not in the picture and does not need to be: the lock screen sets
/// `MPMediaItemPropertyTitle` beside the artwork, the same way every surface in the app states
/// the title beside the well rather than inside it.
@MainActor
public enum PlayerArtworkImage {

    /// How large the picture is rendered.
    ///
    /// One size, and generous: `MPMediaItemArtwork` is asked for a picture at whatever size the
    /// lock screen, Control Centre or a car display wants, and the system scales one bitmap down
    /// far better than it scales a small one up.
    public static let side: CGFloat = 512

    /// The artwork for a publication with no cover of its own.
    ///
    /// - Parameter format: the publication's format, which chooses the glyph and names itself
    ///   under it.
    ///
    /// **Drawn in the dark palette rather than the reader's, deliberately.** The system shows
    /// this over its own material, and an image that followed the app's appearance would have to
    /// be re-rendered and re-published every time a reader flipped it mid-book — a picture on a
    /// lock screen is not a surface that participates in the app's theme.
    ///
    /// - Returns: PNG bytes, or `nil` where the renderer produced nothing — a `nil` the caller
    ///   publishes as *no artwork* rather than as a blank square.
    public static func png(format: PublicationFormat) -> Data? {
        #if canImport(UIKit)
        let renderer = ImageRenderer(
            content: PlayerArtwork(format: format)
                .frame(width: side, height: side)
                .environment(\.theme, Theme(palette: .dark))
        )
        // One point to one pixel: `side` is already in pixels for this purpose, and the default
        // scale would produce a 1536-square image for a 512-square request.
        renderer.scale = 1
        return renderer.uiImage?.pngData()
        #else
        nil
        #endif
    }
}

File: apps/android/app/src/main/kotlin/app/storyarc/PlayerArtwork.kt
package app.storyarc

import android.graphics.Bitmap
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.runtime.withFrameNanos
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawWithContent
import androidx.compose.ui.graphics.asAndroidBitmap
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.graphics.layer.drawLayer
import androidx.compose.ui.graphics.rememberGraphicsLayer
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.unit.dp
import app.storyarc.core.designsystem.cover.CoverlessWell
import app.storyarc.core.designsystem.theme.LocalStoryArcPalette
import app.storyarc.core.designsystem.tokens.StoryArcRadius
import app.storyarc.core.model.Publication

/**
 * How large the cover is asked for. The same 512 the shelves ask for, so a book whose cover
 * the library has already drawn is answered from the cache rather than decoded again.
 */
private const val ARTWORK_PIXELS = 512

/** How wide the artwork may be: iOS's `PlayerArtwork` is 320 pt, and the two players agree. */
private val ARTWORK_MAX_WIDTH = 320.dp

/**
 * The player's artwork: the cover, or the well every other surface draws when there is none.
 *
 * `audio-playback`, *The full player*: it "shows the cover, the publication, the chapter …".
 * *A publication with no cover*: the player "draws the same coverless treatment every other
 * surface draws … rather than a treatment of the player's own", and "the system's own media
 * controls are given that same artwork rather than a second one". Until this file the Android
 * player drew nothing above its chapter name — not a cover, not a well, not a glyph — which
 * `docs/designs/screenshots/android-player-2026-09-04/android-player-full.png` shows, and the
 * design review of 2026-09-06 named as its finding 8.
 *
 * **The well is [CoverlessWell] and not a shape of this file's own.** It is the composable the
 * library shelf, Downloads, Home's cards and the series shelf all draw, in `:core:designsystem`
 * so that `:app` can call it; a second implementation here would be the drift its own header
 * warns about. It takes this platform's two parameters, the title and the format, exactly as
 * `DownloadsParts.kt` hands them over — the format-symbol treatment iOS's well adopted is
 * `audiobooks-and-playback` §4.4b's `:core:designsystem` work, recorded there, and is not
 * decided by a player.
 *
 * **A square rather than 2:3, for the reason iOS's `PlayerArtwork` gives.** An audiobook's
 * artwork is square everywhere a listener has seen one, and the picture below is what the
 * shade and a car display are handed. The well fills whatever shape it is given; a cover is
 * letterboxed onto `surfaceSunken` inside the square rather than cropped, which is the rule
 * every cover cell in this app already follows.
 *
 * **Silent to a screen reader, on both branches.** The top bar states the title and the
 * chapter list states the parts; an `Image` with no description and a well that clears its own
 * semantics add nothing, so TalkBack names the book once. `PlayerSemanticsTest` pins that.
 *
 * **The same picture goes to the media session**, when a sink is given: the drawn pixels are
 * recorded into a graphics layer and handed to [onArtwork] once the artwork has settled — the
 * cover has arrived, or the answer is that there is none — so the notification and the lock
 * screen show what the player shows rather than a second treatment kept in step by hand. iOS
 * renders the same view to PNG for `MPMediaItemPropertyArtwork` for the same reason. Null in a
 * test, where there is no session to hand anything to and no renderer to hand it from.
 *
 * @param title what the publication is called — the well's stand-in for artwork.
 * @param publication the publication being played, or null when the app does not know it: a
 *   book put back on the air by the system's own carousel after the process died reaches the
 *   player with an id nothing in the app has seen. The well then names no format.
 * @param cover where artwork comes from, as a function rather than a view model, so a test can
 *   hand over a cover or the absence of one in a line. `OnDeviceCover` takes its cover the
 *   same way for the same reason.
 */
@Composable
internal fun PlayerArtwork(
    title: String,
    publication: Publication?,
    cover: suspend (Publication, Int) -> Bitmap?,
    onArtwork: ((Bitmap) -> Unit)?,
    modifier: Modifier = Modifier,
) {
    val palette = LocalStoryArcPalette.current
    var art by remember(publication?.id) { mutableStateOf<Bitmap?>(null) }
    // Whether the cover question has been answered, either way. The picture handed to the
    // session waits for this, or the lock screen would get the well for a book that has a
    // cover arriving a frame later.
    var isResolved by remember(publication?.id) { mutableStateOf(publication == null) }
    LaunchedEffect(publication?.id) {
        val asked = publication ?: return@LaunchedEffect
        art = cover(asked, ARTWORK_PIXELS)
        isResolved = true
    }

    val layer = rememberGraphicsLayer()
    // Recorded only where something will read it. A test composes this with no sink and no
    // renderer worth the name, and a layer nobody reads is a layer that need not exist.
    val recorded = if (onArtwork == null) {
        Modifier
    } else {
        Modifier.drawWithContent {
            layer.record { this@drawWithContent.drawContent() }
            drawLayer(layer)
        }
    }

    Box(modifier = modifier.fillMaxWidth(), contentAlignment = Alignment.Center) {
        Box(
            modifier = Modifier
                .widthIn(max = ARTWORK_MAX_WIDTH)
                .fillMaxWidth()
                .aspectRatio(1f)
                .then(recorded)
                .clip(RoundedCornerShape(StoryArcRadius.lg))
                .background(palette.surfaceSunken),
        ) {
            val drawn = art
            if (drawn != null) {
                Image(
                    bitmap = drawn.asImageBitmap(),
                    // The top bar names the book. A description here would name it twice.
                    contentDescription = null,
                    contentScale = ContentScale.Fit,
                    modifier = Modifier.fillMaxSize(),
                )
            } else {
                CoverlessWell(title = title, format = publication?.format?.displayName)
            }
        }
    }

    if (onArtwork != null) {
        LaunchedEffect(isResolved, art) {
            if (!isResolved) return@LaunchedEffect
            // Two frames: one for the resolved artwork to compose, one for it to be drawn
            // into the layer. Reading the layer before it has recorded anything is an empty
            // picture on the lock screen.
            withFrameNanos { }
            withFrameNanos { }
            val picture = runCatching { layer.toImageBitmap() }.getOrNull() ?: return@LaunchedEffect
            onArtwork(picture.asAndroidBitmap())
        }
    }
}
